ctutil/audit.go

```diff
@@ -0,0 +1,183 @@
+
+// Package ctutil contains utilities for Certificate Transparency.
+package ctutil
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/sha256"
+	"encoding/base64"
+	"errors"
+	"fmt"
+
+	ct "github.com/google/certificate-transparency-go"
+	"github.com/google/certificate-transparency-go/tls"
+	"github.com/google/certificate-transparency-go/x509"
+)
+
+var emptyHash = [32]byte{}
+
+// B64LeafHash does as LeafHash does, but returns the leaf hash base64-encoded.
+// The base64-encoded leaf hash returned by B64LeafHash can be used with the
+// get-proof-by-hash API endpoint of Certificate Transparency Logs.
+func B64LeafHash(chain []*x509.Certificate, sct *ct.SignedCertificateTimestamp, embedded bool) (string, error) {
+	hash, err := LeafHash(chain, sct, embedded)
+	if err != nil {
+		return "", err
+	}
+	return base64.StdEncoding.EncodeToString(hash[:]), nil
+}
+
+// LeafHash calculates the leaf hash of the certificate or precertificate at
+// chain[0] that sct was issued for.
+//
+// sct is required because the SCT timestamp is used to calculate the leaf hash.
+// Leaf hashes are unique to (pre)certificate-SCT pairs.
+//
+// This function can be used with three different types of leaf certificate:
+//   - X.509 Certificate:
+//       If using this function to calculate the leaf hash for a normal X.509
+//       certificate then it is enough to just provide the end entity
+//       certificate in chain. This case assumes that the SCT being provided is
+//       not embedded within the leaf certificate provided, i.e. the certificate
+//       is what was submitted to the Certificate Transparency Log in order to
+//       obtain the SCT.  For this case, set embedded to false.
+//   - Precertificate:
+//       If using this function to calculate the leaf hash for a precertificate
+//       then the issuing certificate must also be provided in chain.  The
+//       precertificate should be at chain[0], and its issuer at chain[1].  For
+//       this case, set embedded to false.
+//   - X.509 Certificate containing the SCT embedded within it:
+//       If using this function to calculate the leaf hash for a certificate
+//       where the SCT provided is embedded within the certificate you
+//       are providing at chain[0], set embedded to true.  LeafHash will
+//       calculate the leaf hash by building the corresponding precertificate.
+//       LeafHash will return an error if the provided SCT cannot be found
+//       embedded within chain[0].  As with the precertificate case, the issuing
+//       certificate must also be provided in chain.  The certificate containing
+//       the embedded SCT should be at chain[0], and its issuer at chain[1].
+//
+// Note: LeafHash doesn't check that the provided SCT verifies for the given
+// chain.  It simply calculates what the leaf hash would be for the given
+// (pre)certificate-SCT pair.
+func LeafHash(chain []*x509.Certificate, sct *ct.SignedCertificateTimestamp, embedded bool) ([32]byte, error) {
```

Comment:
`[sha256.Size]byte` might be better (and seems more common across the codebase)

Reply:
Done here and for the emptyHash at the top of this file.